File: TraaS/src/de/tudresden/sumo/cmd/Person.java
/*   
    Copyright (C) 2017 Mario Krumnow, Dresden University of Technology

    This file is part of TraaS.

    TraaS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License.

    TraaS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with TraaS.  If not, see <http://www.gnu.org/licenses/>.
*/


package de.tudresden.sumo.cmd;
import de.tudresden.sumo.config.Constants;
import de.tudresden.sumo.util.SumoCommand;

/**
 * 
 * @author Mario Krumnow
 *
 */

public class Person {

	//getter methods

	
	/**
	 * Returns the number of all persons in the network.
	 * @return the number of persons in the network
	 */

	public static SumoCommand getIDCount(){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.ID_COUNT, "", Constants.RESPONSE_GET_PERSON_VARIABLE, Constants.TYPE_INTEGER);
	}
	
	
	/**
	 * Returns a list of personIDs of all persons
	 * 
	 * @return list of IDs of all persons
	 */

	public static SumoCommand getIDList(){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.ID_LIST, "", Constants.RESPONSE_GET_PERSON_VARIABLE, Constants.TYPE_STRINGLIST);
	}
	
	/**
	 * If the person is walking, returns the next edge on the persons route (including crossing and walkingareas). If there is no further edge or the person is in another stage, returns the empty string.
	 * 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return next edge
	 */

	public static SumoCommand getNextEdge(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_NEXT_EDGE, personID, Constants.RESPONSE_GET_PERSON_VARIABLE, Constants.TYPE_STRING);
	}
	


	/**
	 * Returns the chosen parameter
	 *
	 *  @param personID a string personIDentifying the person
	 *  @param param a string personIDentifying the parameter
	 *  
	 * @return the specific parameter
	 */

	public static SumoCommand getParameter(String personID, String param){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_PARAMETER, personID, Constants.RESPONSE_GET_PERSON_VARIABLE, Constants.TYPE_STRING);
	}
	
	/**
	 * get speed
	 * 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the speed in m/s
	 */
	
	public static SumoCommand getSpeed(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_SPEED, personID, Constants.RESPONSE_GET_PERSON_VARIABLE, Constants.TYPE_DOUBLE);
	}

	/**
	 * get position
	 * 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the position 2D
	 */
	
	public static SumoCommand getPosition(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_POSITION, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.POSITION_2D);
	}
	
	/**
	 * get position3D
	 * 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the position 3D
	 */
	
	public static SumoCommand getPosition3D(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_POSITION3D, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.POSITION_3D);
	}
	
	/**
	 * get angle
	 * 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the angle
	 */
	
	public static SumoCommand getAngle(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_ANGLE, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_DOUBLE);
	}
	
	/**
	 * get road ID
	 * 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the personID of the road
	 */

	public static SumoCommand getRoadID(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_ROAD_ID, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_STRING);
	}

	/**
	 * get type ID
	 * 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the type ID of the person
	 */
	
	public static SumoCommand getTypeID(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_TYPE, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_STRING);
	}
	
	/**
	 * get lane position 
	 * 	 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the position on the lane
	 */
	
	public static SumoCommand getLanePosition(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_LANEPOSITION, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_DOUBLE);
	}
	
	/**
	 * get color 
	 * 	 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the color of the person in the GUI
	 */
	
	public static SumoCommand getColor(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_COLOR, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_COLOR);
	}
	
	
	/**
	 * get person number
	 * 	 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the person number
	 */
	
	public static SumoCommand getPersonNumber(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_PERSON_NUMBER, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_INTEGER);
	}
	
	/**
	 * get length
	 * 	 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the length of the person
	 */
	
	public static SumoCommand getLength(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_LENGTH, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_DOUBLE);
	}
	
	/**
	 * get waiting time
	 * 	 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the waiting time of the person
	 */
	
	
	public static SumoCommand getWaitingTime(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_WAITING_TIME, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_DOUBLE);
	}
	
	/**
	 * get wpersonIDth
	 * 	 
	 * @param personID
	 *            a string identifying the person
	 * @return the width of the person
	 */
	
	public static SumoCommand getWidth(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_WIDTH, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_DOUBLE);
	}
	
	/**
	 * get minGap
	 * 	 
	 * @param personID
	 *            a string personIDentifying the person
	 * @return the value for the minimum Ggap of the person
	 */
	
	public static SumoCommand getMinGap(String personID){
		return new SumoCommand(Constants.CMD_GET_PERSON_VARIABLE, Constants.VAR_MINGAP, personID, Constants.RESPONSE_GET_PERSON_VARIABLE,Constants.TYPE_DOUBLE);
	}
	
}

File: sumo/tools/traci/person.py
# -*- coding: utf-8 -*-
"""
@file    person.py
@author  Jakob Erdmann
@date    2015-02-06
@version $Id$

Python implementation of the TraCI interface.

SUMO, Simulation of Urban MObility; see http://sumo.dlr.de/
Copyright (C) 2011-2015 DLR (http://www.dlr.de/) and contributors

This file is part of SUMO.
SUMO is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.
"""
import struct
import traci
import traci.constants as tc


_RETURN_VALUE_FUNC = {tc.ID_LIST:             traci.Storage.readStringList,
                      tc.ID_COUNT:            traci.Storage.readInt,
                      tc.VAR_SPEED:           traci.Storage.readDouble,
                      tc.VAR_POSITION: lambda result: result.read("!dd"),
                      tc.VAR_ANGLE:           traci.Storage.readDouble,
                      tc.VAR_ROAD_ID:         traci.Storage.readString,
                      tc.VAR_TYPE:            traci.Storage.readString,
                      tc.VAR_ROUTE_ID:        traci.Storage.readString,
                      tc.VAR_COLOR: lambda result: result.read("!BBBB"),
                      tc.VAR_LANEPOSITION:    traci.Storage.readDouble,
                      tc.VAR_LENGTH:          traci.Storage.readDouble,
                      tc.VAR_WAITING_TIME:    traci.Storage.readDouble,
                      tc.VAR_WIDTH:           traci.Storage.readDouble,
                      tc.VAR_MINGAP:          traci.Storage.readDouble,
                      }

subscriptionResults = traci.SubscriptionResults(_RETURN_VALUE_FUNC)


def _getUniversal(varID, personID):
    result = traci._sendReadOneStringCmd(
        tc.CMD_GET_PERSON_VARIABLE, varID, personID)
    return _RETURN_VALUE_FUNC[varID](result)


def getIDList():
    """getIDList() -> list(string)

    Returns a list of ids of all persons currently running within the scenario.
    """
    return _getUniversal(tc.ID_LIST, "")


def getIDCount():
    """getIDCount() -> integer

    Returns the number of person in the network.
    """
    return _getUniversal(tc.ID_COUNT, "")


def getSpeed(personID):
    """getSpeed(string) -> double

    Returns the speed in m/s of the named person within the last step.
    """
    return _getUniversal(tc.VAR_SPEED, personID)


def getPosition(personID):
    """getPosition(string) -> (double, double)

    Returns the position of the named person within the last step [m,m].
    """
    return _getUniversal(tc.VAR_POSITION, personID)


def getPosition3D(personID):
    """getPosition(string) -> (double, double)

    Returns the position of the named person within the last step [m,m,m].
    """
    return _getUniversal(tc.VAR_POSITION3D, personID)


def getAngle(personID):
    """getAngle(string) -> double

    Returns the angle in degrees of the named person within the last step. 
    """
    return _getUniversal(tc.VAR_ANGLE, personID)


def getRoadID(personID):
    """getRoadID(string) -> string

    Returns the id of the edge the named person was at within the last step.
    """
    return _getUniversal(tc.VAR_ROAD_ID, personID)


def getTypeID(personID):
    """getTypeID(string) -> string

    Returns the id of the type of the named person.
    """
    return _getUniversal(tc.VAR_TYPE, personID)


def getLanePosition(personID):
    """getLanePosition(string) -> double

    The position of the person along the lane measured in m.
    """
    return _getUniversal(tc.VAR_LANEPOSITION, personID)


def getColor(personID):
    """getColor(string) -> (integer, integer, integer, integer)

    Returns the person's rgba color.
    """
    return _getUniversal(tc.VAR_COLOR, personID)


def getPersonNumber(personID):
    """getPersonNumber(string) -> integer

    .
    """
    return _getUniversal(tc.VAR_PERSON_NUMBER, personID)


def getLength(personID):
    """getLength(string) -> double

    Returns the length in m of the given person.
    """
    return _getUniversal(tc.VAR_LENGTH, personID)


def getWaitingTime(personID):
    """getWaitingTime() -> double
    The waiting time of a person is defined as the time (in seconds) spent with a
    speed below 0.1m/s since the last time it was faster than 0.1m/s.
    (basically, the waiting time of a person is reset to 0 every time it moves). 
    """
    return _getUniversal(tc.VAR_WAITING_TIME, personID)


def getWidth(personID):
    """getWidth(string) -> double

    Returns the width in m of this person.
    """
    return _getUniversal(tc.VAR_WIDTH, personID)


def getMinGap(personID):
    """getMinGap(string) -> double

    Returns the offset (gap to front person if halting) of this person.
    """
    return _getUniversal(tc.VAR_MINGAP, personID)


def subscribe(personID, varIDs=(tc.VAR_ROAD_ID, tc.VAR_LANEPOSITION), begin=0, end=2**31 - 1):
    """subscribe(string, list(integer), double, double) -> None

    Subscribe to one or more person values for the given interval.
    """
    traci._subscribe(
        tc.CMD_SUBSCRIBE_PERSON_VARIABLE, begin, end, personID, varIDs)


def getSubscriptionResults(personID=None):
    """getSubscriptionResults(string) -> dict(integer: <value_type>)

    Returns the subscription results for the last time step and the given person.
    If no person id is given, all subscription results are returned in a dict.
    If the person id is unknown or the subscription did for any reason return no data,
    'None' is returned.
    It is not possible to retrieve older subscription results than the ones
    from the last time step.
    """
    return subscriptionResults.get(personID)


def subscribeContext(personID, domain, dist, varIDs=(tc.VAR_ROAD_ID, tc.VAR_LANEPOSITION), begin=0, end=2**31 - 1):
    traci._subscribeContext(
        tc.CMD_SUBSCRIBE_PERSON_CONTEXT, begin, end, personID, domain, dist, varIDs)


def getContextSubscriptionResults(personID=None):
    return subscriptionResults.getContext(personID)


getParameter, setParameter = traci.getParameterAccessors(
    tc.CMD_GET_PERSON_VARIABLE, tc.CMD_SET_PERSON_VARIABLE)
